### PR TITLE
Establish basic color patterns

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR establishes some very basic structure around managing colors within this package's internal SASS. This is part of ongoing work to start to establish some foundational bits and bobs of our design system—not creating a new system, but describing the existing one.

Changes relevant to package users: additional page visible in the pattern library (Colors) with some basic color swatches, and a  sub-heading changed from _Common Patterns_ to _Foundations_. No externally-consumable SASS or CSS is affected by these changes.

These changes are internal to the package:

* Establishes a 0 - 9 greyscale from existing color values. Internally some variable names changed, but no colors changed—just their references. Previously we had a 0-7 greyscale range, with some intermediate `grey-mid` and `grey-semi` values. One had to remember (or forget, every time) whether `grey-semi` was lighter or darker than `grey-4`. No longer. Values are lightest at 0 and darkest at 9.
* Standardizes naming for color variables
* Generates utility classes for foreground and background colors

There's a lot more to do here, but I wanted to keep things as simple as possible. These changes were motivated by some needs for `Dialog`—it felt like the right time to do this.

Screenshot from new pattern-library page:

image

There's a lot to do here (I left myself some TODOs, even), but it's a start.

### CODE REVIEW

File: src/pattern-library/components/PlaygroundApp.js [hunk 1]
```diff
@@ -109,7 +115,7 @@ export default function PlaygroundApp({
             <SvgIcon name="logo" />
           </a>
         </div>
-        <h2>Common Patterns</h2>
+        <h2>Foundations</h2>
```
Comment (lyzadanger) [inline src/pattern-library/components/PlaygroundApp.js:118]: This just seemed more accurate.

File: styles/components/buttons/_config.scss [hunk 2]
```diff
@@ -8,62 +8,62 @@ $color-g1: var.$color-grey-1 !default;
 $color-g2: var.$color-grey-2 !default;
 $color-g3: var.$color-grey-3 !default;
 $color-g4: var.$color-grey-4 !default;
-$color-g6: var.$color-grey-6 !default;
+$color-g5: var.$color-grey-5 !default;
```
Comment (lyzadanger) [inline styles/components/buttons/_config.scss:11]: The changes in this module are just updating internal variable names. No visible changes—resulting colors are the same.

File: styles/util/_colors.scss [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+// Generate a foreground-color utility class for each color
+// and an override foreground-color class for same, e.g.
+// `.hyp-u-color--grey-0` and `.hyp-!-u-color--grey-0`
+@each $color-name, $color-value in $all-colors {
+  .hyp-u-color--#{$color-name} {
+    color: $color-value;
+  }
+
+  .hyp-\!-u-color--#{$color-name} {
```
Comment (lyzadanger) [inline styles/util/_colors.scss:29]: I will talk to the group at our next sync about the motivation here; there is an example on the new colors pattern-library page that illustrates a use case.

File: styles/variables/_colors.scss [hunk 4]
```diff
@@ -1,47 +1,28 @@
...
 $grey-3: #dbdbdb;
-
 $grey-4: #a6a6a6;
-
-// Interim color variable for migration purposes, as the step between `$grey-4`
-// and `$grey-5` is large. Represents `base-semi` in proposed future palette,
-// minus blue tint.
-$grey-semi: #9c9c9c;
-
-// This is the lightest grey admissible on a white, $grey-0 or $grey-1
```
Comment (lyzadanger) [inline styles/variables/_colors.scss:21]: This information needs to be surfaced in the pattern library (and will be).

File: src/pattern-library/components/patterns/SharedColorPatterns.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      <Pattern title="Color swatches">
+        <PatternExamples>{backgroundExamples}</PatternExamples>
+      </Pattern>
+
+      <Pattern title="Overriding background colors: example">
+        <PatternExamples>
+          <PatternExample details="Background utility class without override: panel specificity wins">
+            <div className="hyp-panel hyp-u-bg-color--grey-2">
+              This is a <code>panel</code> with an applied utility class
+              <code>.hyp-u-bg-color--grey-2</code>.
```
Comment (esanzgar) [inline src/pattern-library/components/patterns/SharedColorPatterns.js:51]:
```suggestion
              This is a <code>panel</code> with an applied utility class
              <code>.hyp-u-bg-color--grey-2</code> that is ignored.
```

**Review by esanzgar — APPROVED** [review]
Looks great.

I personally can't distinguish between these two colours:
$grey-4: #a6a6a6;
$grey-5: #9c9c9c;

**Comment by lyzadanger** [comment]
@esanzgar 

> I personally can't distinguish between these two colours:
> $grey-4: #a6a6a6;
> $grey-5: #9c9c9c;

Totally. First we have to _describe_ the system...then maybe we can fix it :)

File: src/pattern-library/components/patterns/SharedColorPatterns.js (hunk 5 above)
Comment (lyzadanger) [reply src/pattern-library/components/patterns/SharedColorPatterns.js]: Thanks, @esanzgar ! I ended up slightly rewriting both descriptions here. The examples are a bit fugly; I hope to improve them over time.

**Pull request merged**